python/ray/serve/controller.py

```diff
@@ -151,40 +149,40 @@ async def __init__(self, instance_name, http_node_id, http_proxy_host,
             asyncio.get_event_loop().create_task(
                 self._recover_from_checkpoint(checkpoint))
 
-    def _get_or_start_http_proxy(self, node_id, host, port):
+    def _get_or_start_routers(self, host, port):
         """Get the HTTP proxy belonging to this serve instance.
 
         If the HTTP proxy does not already exist, it will be started.
         """
-        proxy_name = format_actor_name(SERVE_PROXY_NAME, self.instance_name)
-        try:
-            self.http_proxy = ray.get_actor(proxy_name)
-        except ValueError:
-            logger.info(
-                "Starting HTTP proxy with name '{}' on node '{}'".format(
-                    proxy_name, node_id))
-            self.http_proxy = HTTPProxyActor.options(
-                name=proxy_name,
-                max_concurrency=ASYNC_CONCURRENCY,
-                max_restarts=-1,
-                max_task_retries=-1,
-                resources={
-                    node_id: 0.01
-                },
-            ).remote(
-                host, port, instance_name=self.instance_name)
-
-        # Since router is a merged with HTTP proxy actor, the router will be
-        # proxied via the HTTP actor. Even though the two variable names are
-        # pointing to the same object, their semantic differences make the code
-        # more readable. (e.g. http_proxy.set_route_table, router.add_worker)
-        self.router = self.http_proxy
-
-    def get_http_proxy(self):
+        # TODO(simon): We don't handle nodes being added/removed. To do that,
+        # we should implement some sort of control loop in master actor.
+        for i, node_id in enumerate(ray.state.node_ids()):
+            proxy_name = format_actor_name(SERVE_PROXY_NAME,
+                                           self.instance_name)
```

Comment:
We should probably use the node ID in the name here instead of the index - not sure if the ordering in `ray.state.node_ids` is deterministic which may cause problems in the future. Naming it with node ID will probably also be more convenient (e.g., for tests).

Reply:
made it to "{SERVE_PROXY_NAME}-{self.instance_name}-{node_id}-{index}" so we allow multiple servers per node.